Issue #76 - Update opencascade to 7.6.2 (fork)

hfiguiere opened the pull request:
- Also update IfcOpenShell 0.7.0a0 for compatibility
- Add CGAL (and mpfr) as required for IfcOpenShell
- Added boost::chrono and boost::atomic as a CGAL dependency

hfiguiere committed:
[beta] Issue #76 - Update opencascade to 7.6.2 (fork)

- Also update IfcOpenShell 0.7.0a2 for compatibility
- Add CGAL 5.3.1 (and mpfr) as required for IfcOpenShell
- Update boost to 1.71 (for IfcOpenShell)
- Added boost::chrono and boost::atomic as a CGAL dependency

adrianinsaval commented:
I'll ask blobfish if he can provide a 7.5.3 tag with his patches, there were many issues with step export and import in the early 7.5 series

hfiguiere commented on org.freecadweb.FreeCAD.yaml:
```diff
@@ -482,13 +499,15 @@ modules:
-        sha256: c948327d5ef11e7322244489e5d72f9651472e9f571ae3cf1ea393aef331abb9
+      - type: git
+        url: https://gitlab.com/blobfish/occt
+        tag: V7_6_2_BF
```
Note for later: this is a branch, not a tag. Need to pick a commit.

hfiguiere commented:
Upstream issue for 7.5.x patches https://gitlab.com/blobfish/occt/-/issues/1

adrianinsaval commented:
according to one of the freecad maintainers blobfish said he wouldn't provide a 7.5.3 branch so probably better to just go to 7.6.2 once we figure out ifcopenshell

hfiguiere commented:
bot, build

adrianinsaval commented:
ifcopenshell 0.7 uses submodules and the autogenerated github archive doesn't include them, you need a recursive git clone

hfiguiere commented:
JFC. why do people do that?

adrianinsaval commented:
Submodules can be useful when using third party code, github not including submodules in the archive escapes my understanding.

hfiguiere commented:
I know what submodules are for.

adrianinsaval commented:
Looks like you forgot to change the source type to git. My sympathies and thanks to you, all this is probably frustrating.

hfiguiere commented:
I usually build locally. I didn't even try, which is all my fault.

Otherwise I'd have seen it
```
(flatpak-builder:2047505): flatpak-builder-WARNING **: 17:31:55.999: Unknown property tag for type BuilderSourceArchive

(flatpak-builder:2047505): flatpak-builder-WARNING **: 17:31:55.999: Unknown property commit for type BuilderSourceArchive
```

hfiguiere commented:
JFC. This is not a git submodule.

adrianinsaval commented:
> Failed to download sources: module IfcOpenShell: Not a gitlink tree: src/ifcopenshell-python/ifcopenshell/Sample-BIM-Files

That's a very strange error, there is no submodule with that name on that path but there is on` src/ifcopenshell-python/test/`
I don't get any errors running `git clone  --recurse-submodules --branch v0.7.0a0 https://github.com/IfcOpenShell/IfcOpenShell.git`

adrianinsaval commented:
@aothms any clues what is happening here?

aothms commented:
I don't know anything abt flatpak, but it appears it's scanning submodules by parsing

    git show 0f2a545cbb991de390a7e5d9c0978e437e9c63c0:.gitmodules

And this tripped over a wrong entry that `git submodule foreach` (or indeed `git clone --recursive`) was more permissive abt, so I never noticed.

Should be fixed:

https://github.com/IfcOpenShell/IfcOpenShell/commit/524fa58f0501e86276cd69957f1aa4991a503306

New tag:

https://github.com/IfcOpenShell/IfcOpenShell/tree/v0.7.0a1

hfiguiere commented:
Unsure about the patch for the build system, but looking up static libraries in the wrong place is definitely a double wrong.

Also on my machine, the build process get killed building Ifc4x2.cpp. I suspect it's running out of RAM. (I have 16GB)

hfiguiere commented:
So which version of CGAL to I need? Dependency checking doesn't say, nor the README.

aothms commented:
> So which version of CGAL to I need? Dependency checking doesn't say, nor the README.

Honestly I think most versions would work. It's just a matter of being configured header-only or not, for which the default switched I think in v5.

> Also on my machine, the build process get killed building Ifc4x2.cpp. I suspect it's running out of RAM. (I have 16GB)

I didn't see from your conf how many simultaneous build procs you have, but schema-specific files are indeed huge and can easily take 4g RAM to compile.

> Unsure about the patch for the build system, but looking up static libraries in the wrong place is definitely a double wrong.

In IfcOpenShell, we don't have the breadth and depth of contributor experience that e.g freecad has. Things like the build and configure scripts don't always live up to expectations.

Rather than maintaining the patch here, contributions are very very welcome upstream :) Most IfcOpenShell users are rather novices with compilation et al. and hence we kind of default on linking everything statically, which is indeed a bit unconventional on the nixes.

I don't think the patch is necessary when you specify the libraries required for hdf5 directly as configuration options like here https://github.com/IfcOpenShell/IfcOpenShell/issues/1759#issuecomment-955709694 But yeah, it's messy and kind of eliminates the point of having a cmake script.

hfiguiere commented:
> I don't think the patch is necessary when you specify the libraries required for hdf5 directly as configuration options like here [IfcOpenShell/IfcOpenShell#1759 (comment)](https://github.com/IfcOpenShell/IfcOpenShell/issues/1759#issuecomment-955709694) But yeah, it's messy and kind of eliminates the point of having a cmake script.

It does. It tries to find `/app/lib/libz.a` which is double wrong. Because there is no `.a` and it is not in the same prefix as hdf5. Also doesn't hdf5 come with a `.pc` ?

Note that sure submitting PR, but when I don't even know if that works, there is no point.

aothms commented:
I meant the patch is probably not necessary when you use the equivalent of `HDF5_LIBRARIES=/usr/lib/libhdf5_cpp.so;/usr/lib/libhdf5.so;/usr/lib/libz;/usr/lib/libSZ.so;/usr/lib/libaec.so` instead of HDF5_LIBRARY_**DIR**, but again, I agree it's sloppy.

hfiguiere commented:
looks like a boost build failure.

hfiguiere commented:
maybe I should update to 1.71....

hfiguiere commented:
not sure how to test IfcOpenShell properly, or the new things from opencascade.

hfiguiere commented on org.freecadweb.FreeCAD.yaml:
```diff
@@ -540,10 +575,19 @@ modules:
+        options:
+          - -d
+          - src/svgfill
+        path: patches/IfcOpenShell-svgfill.patch
```
just a not, I'm not sure if this patch is needed anymore given what the top error was. I doubt it hurts.

hfiguiere commented on patches/IfcOpenShell-hdf5.patch:
```diff
@@ -0,0 +1,20 @@
+-            "${HDF5_LIBRARY_DIR}/libz${zlib_post}${debug_postfix}.${lib_ext}"
+-            "${HDF5_LIBRARY_DIR}/libsz${debug_postfix}.${lib_ext}"
+-            "${HDF5_LIBRARY_DIR}/libaec${debug_postfix}.${lib_ext}"
++            "/usr/lib/x86_64-linux-gnu/libz${zlib_post}${debug_postfix}.${lib_ext}"
```
FIXME: this is bad for `aarch64` support.

adrianinsaval commented:
We could merge this to beta and I'll ask people on the forum to test ifcopenshell, how do I pass arguments to the command when running the flatpak? Would be a good idea to run the freecad tests. I'll search later for the cases that crashed the old OCC and test with this.

hfiguiere commented:
I haven't tested it at all yet, so I won't merge before that. It's not because it builds that it works :-)

adrianinsaval commented:
If you can run `FreeCAD --run-test 0` I don't have a Linux pc on hand right now

adrianinsaval commented:
I ran the console tests on WSL now (FreeCAD -c --run-test 0) and it passed

hfiguiere commented:
I have a certain number of test failures though.

Also do we need `IfcGeomServer` and `IfcConvert`

Alone they represent 200+MB (the package went from 960MB to 2.2GB)

hfiguiere commented:
According to https://wiki.freecadweb.org/IfcOpenShell

> The "Blender", "IfcConvert", and "IfcGeomServer" packages are not needed for FreeCAD usage.

So the answer is no.

adrianinsaval commented:
1.2gb extra?? Where is all that coming from.
Can you post your test failures? Note that it's normal that FreeCAD raises warnings an errors during the test, that is part of what gets tested, at the end of the log it should say if all test passed or how many failed.

hfiguiere commented:
some are related to openscad which should work but has a different setup. I'll post them after shrinking that build.

hfiguiere commented:
Also it seem we need open collada.

adrianinsaval commented:
I think what is used is pycollada, not sure if related or a wrapper for opencollada or what. For reference here are the conda packages the appimage uses:
https://github.com/looooo/freecad-feedstock/blob/7b18e86bac28aa5103c50bed101eebdbd39f9659/recipe/meta.yaml#L22
https://github.com/FreeCAD/FreeCAD-Bundle/blob/master/conda/linux/create_bundle.sh#L3-L11

hfiguiere commented:
We have pycollada already, and IfcOpenShell is built with Collada support disabled.

hfiguiere commented:
It was disabled like that before the update.

hfiguiere commented:
The error:
```
======================================================================
FAIL: test01 (PathTests.TestPathToolController.TestPathToolController)
Verify ToolController template roundtrip.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/app/freecad/Mod/Path/PathTests/TestPathToolController.py", line 102, in test01
    self.assertRoughly(tc0.Tool.Diameter, tc1.Tool.Diameter)
  File "/app/freecad/Mod/Path/PathTests/PathTestUtils.py", line 37, in assertRoughly
    self.assertTrue(math.fabs(f1 - f2) < error, "%f != %f" % (f1, f2))
AssertionError: False is not true : 1.750000 != 3.175000

----------------------------------------------------------------------
Ran 796 tests in 192.922s

FAILED (failures=1, errors=39)
```

(I skipped the 39 errors)

adrianinsaval commented:
I isntalled the flatpak with
`flatpak install --user https://dl.flathub.org/build-repo/91875/org.freecadweb.FreeCAD.flatpakref`
and when running the tests with
`flatpak run org.freecadweb.FreeCAD --run-test 0`
I get this:

> Ran 772 tests in 180.260s
> 
> OK
> System exit

adrianinsaval commented:
I tried exporting and then importing a simple arch wall as ifc and it worked without issues

hfiguiere commented:
there must be something different here. Weird.

adrianinsaval commented:
I now tested the file from this bug report: https://tracker.freecad.org/view.php?id=4470 and following the instructions it crashes the appimage and an old snap package that I have that uses occt 7.6.2 without the blobfish patches, this flatpak doesn't crash.
The current flatpak from flathub-beta also crashes.

hfiguiere commented:
So I actually installed the test build above and it is 1.3GB. If we remove the 200MB binaries we don't need, we have just grown 150MB. I'll check more what happening with the next build.

hfiguiere committed:
[beta] Issue #76 - Disable IfcConvert and IfcGeomServ

- Shrink down 200MB (not having IfcConvert and IfcGeomServ binaries)
- Remove patch that is actually not needed.

adrianinsaval commented:
Thanks for all the effort!

luzpaz commented:
Indeed! +1
